### PR TITLE
fix logpdf shape of uniform distribution

### PR DESCRIPTION
**Pull request by fehiepsi**

### CODE REVIEW

**Comment by hawkinsp** [comment]
Looks good! Thanks for the PR.

Would you mind also adding a test to `scipy_stats_test.py` to make sure this stays working in the future? Thanks!

**Comment by hawkinsp** [comment]
Actually it looks like some of the existing tests are failing with this change, see the travis CI results.

**Comment by fehiepsi** [comment]
Yeah, let me fix it. :)

File: jax/scipy/stats/uniform.py [hunk 1]
```diff
@@ -20,14 +20,15 @@
...
-from ...numpy.lax_numpy import _promote_args_like, _wraps, where, inf, logical_or
+from ...numpy.lax_numpy import _promote_args_like, _wraps, where, inf, logical_or, broadcast_to
 
 
 @_wraps(osp_stats.uniform.logpdf)
 def logpdf(x, loc=0, scale=1):
   x, loc, scale = _promote_args_like(osp_stats.uniform.logpdf, x, loc, scale)
   fill_value = lax.neg(lax.log(scale))
-  log_probs = lax.broadcast(fill_value, onp.shape(x))
+  log_probs_shape = lax.broadcast_shapes(*[onp.shape(x), onp.shape(loc), onp.shape(scale)])
```
Comment (hawkinsp) [inline jax/scipy/stats/uniform.py]: *[...] is equivalent to just passing the values as arguments directly.

File: tests/scipy_stats_test.py [hunk 2]
```diff
@@ -174,7 +174,12 @@ def args_maker():
       x, loc, scale = map(rng, shapes, dtypes)
       return [x, loc, onp.abs(scale)]
 
+    def args_maker2():
+      _, loc, scale = map(rng, shapes, dtypes)
+      return [0.5, loc, onp.abs(scale)]
+
     self._CheckAgainstNumpy(scipy_fun, lax_fun, args_maker, check_dtypes=True)
+    self._CheckAgainstNumpy(scipy_fun, lax_fun, args_maker2, check_dtypes=True)
```
Comment (hawkinsp) [inline tests/scipy_stats_test.py]: Is the intent to test a scalar here, as opposed to an array that has 0 dimensions? It turns out there's a magic value you can pass as the shape to get the RNG to produce a Numpy scalar:
https://github.com/google/jax/blob/master/tests/lax_numpy_test.py#L44

Does that fulfill your needs?

**Review by hawkinsp — COMMENTED** [review]
Looks almost perfect now, two more small things.

**Comment by hawkinsp** [comment]
(Incidentally, I see you're located in Princeton NJ — so am I! If you'd like to chat about JAX sometime, I'd be happy to do so.)

File: jax/scipy/stats/uniform.py (hunk 1 above)
Comment (fehiepsi) [reply jax/scipy/stats/uniform.py]: Thanks @hawkinsp !

File: tests/scipy_stats_test.py (hunk 2 above)
Comment (fehiepsi) [reply tests/scipy_stats_test.py]: Sorry, my bad. I have interpreted the wrapper genNamedParametersNArgs wrongly. My intend was to make the shapes for `x`, `loc`, `scale` differently but `genNamedParametersNArgs` already did it. So the test which I added was redundant and did not capture the error in uniform implementation. I have replaced it with a new test which explicitly captures the shape error.

**Comment by fehiepsi** [comment]
(@hawkinsp Sure, it'd be great to chat with you! Should I contact through your Stanford email address?)

File: jax/scipy/stats/uniform.py [hunk 3]
```diff
@@ -20,14 +20,15 @@
...
 from ... import lax
-from ...numpy.lax_numpy import _promote_args_like, _wraps, where, inf, logical_or
+from ...numpy.lax_numpy import _promote_args_like, _wraps, where, inf, logical_or, broadcast_to
 
 
 @_wraps(osp_stats.uniform.logpdf)
 def logpdf(x, loc=0, scale=1):
   x, loc, scale = _promote_args_like(osp_stats.uniform.logpdf, x, loc, scale)
   fill_value = lax.neg(lax.log(scale))
-  log_probs = lax.broadcast(fill_value, onp.shape(x))
```
Comment (hawkinsp) [inline jax/scipy/stats/uniform.py]: Sorry for not noticing this before, but I think there's a simpler fix.

Just set `log_probs = lax.neg(lax.log(scale))`, and don't have any additional broadcasting here.

The reason this should work is because `_promote_args_like` also promotes the shape by making all of the arguments of equal rank. And all the other operators here broadcast too. So we shouldn't need any additional broadcasting.

File: tests/scipy_stats_test.py [hunk 4]
```diff
@@ -176,6 +177,7 @@ def args_maker():
 
     self._CheckAgainstNumpy(scipy_fun, lax_fun, args_maker, check_dtypes=True)
     self._CompileAndCheck(lax_fun, args_maker, check_dtypes=True)
+    assert lax_fun(*args_maker()).shape == lax.broadcast_shapes(*shapes)
```
Comment (hawkinsp) [inline tests/scipy_stats_test.py]: I think you can revert the change to the test. The `_CheckAgainstNumpy` should be catching the mismatched shape but is not. I'm going to fix that part separately.

**Comment by hawkinsp** [comment]
> (@hawkinsp Sure, it'd be great to chat with you! Should I contact through your Stanford email address?)

Sure, that still works (hawkinsp at cs dot stanford dot edu).

**Comment by hawkinsp** [comment]
PR #513 means the existing tests catch the bug this PR fixes. So if you submit the fix, I can submit the improved test case. Thanks!

File: jax/scipy/stats/uniform.py (hunk 3 above)
Comment (fehiepsi) [reply jax/scipy/stats/uniform.py]: Nice observation!

File: tests/scipy_stats_test.py (hunk 4 above)
Comment (fehiepsi) [reply tests/scipy_stats_test.py]: Sure, thank you!

**Comment by fehiepsi** [comment]
@hawkinsp I have reverted the changes. Thanks for your suggestion!

**Review by hawkinsp — APPROVED** [review]
Thanks for the PR!
